### PR TITLE
rework staging definitions

### PR DESCRIPTION
**Pull request by odinuv**
https://keboola.atlassian.net/browse/PS-1628

### CODE REVIEW

**Comment by odinuv** [comment]
Guys please look at https://github.com/keboola/input-mapping/pull/78/files#diff-166cb219a4b00719b5b91221d79af4635d2c5bb2ab02d2fbd0e93f243ec4794aR75 - that's the only basically working thing and tell me whether it makes sense (before i try to do the rest) , we could he a little call if you want.

The idea behind addStagingCapability is that this will be called in runner before actually loading data, depending on the project info (i.e. if projectHasRedshift then addStagingCapability(redshiftWorkspace, ...))

File: tests/Functional/DownloadFilesAbsWorkspaceTest.php [hunk 1]
```diff
@@ -109,26 +133,26 @@ public function testAbsReadFiles()
...
         );
 
         self::assertEquals("test", stream_get_contents($blobResult1->getContentStream()));
         self::assertEquals("test", stream_get_contents($blobResult2->getContentStream()));
 
         $adapter = new Adapter();
-        $manifest1 = $adapter->readFromFile($root . "/download/" . $id1 . "_upload.manifest");
-        $manifest2 = $adapter->readFromFile($root . "/download/" . $id2 . "_upload.manifest");
+        $manifest1 = $adapter->readFromFile($root . "/data/in/files/" . $id1 . "_upload.manifest");
+        $manifest2 = $adapter->readFromFile($root . "/data/in/files/" . $id2 . "_upload.manifest");
```
Comment (pivnicek) [inline tests/Functional/DownloadFilesAbsWorkspaceTest.php:155]: so the `metadataStorage` for the abs-workspace will be local disk?

Comment (pivnicek) [reply tests/Functional/DownloadFilesAbsWorkspaceTest.php:155]: sorry, I see that the mock is returning the tmp dir for the getPath method

**Review by pivnicek — COMMENTED** [review]
wow, heavy duty changes :-).  I like the direction, especially the consolidation of the strategies (I didn't like how files and tables had their own separate abstractions) so this is a great improvement.  Also separating the metadata storage from the data storage is nice too.  One thing I didn't quite understand until reading a few times (and maybe still don't :-) ) is the Fulfilment class and what it brings (along with the addStagingCapability thing). I see how the mock is using it, but in general usage these capabilities will all be predefined by the strategy map, no?

File: tests/Functional/DownloadFilesAbsWorkspaceTest.php (hunk 1 above)
Comment (odinuv) [reply tests/Functional/DownloadFilesAbsWorkspaceTest.php:155]: in the input mapping yes it's the local disk the moment - that's the status quo

**Comment by odinuv** [comment]
The fulfillment class is there because there are no (especially workspace) staging implementations in input-mapping library. That means that the e.g. ABS-workspace can't be used (because there is no way to get it) - unless the staging requirement is somehow fulfilled - in case of the tests by a mock

**Comment by pivnicek** [comment]
ah, right, that makes sense.  So once the workspace is created it can "fulfill" those capabilities.  I see.

**Comment by odinuv** [comment]
I'm in for better names :)

File: src/Staging/StrategyFactory.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        $this->clientWrapper = $clientWrapper;
+        $this->logger = $logger;
+        $this->format = $format;
+    }
+
+    /**
+     * @return Definition[]
+     */
+    public function getStrategyMap()
```
Comment (pivnicek) [inline src/Staging/StrategyFactory.php:60]: Maybe we can change the `Definition` name (which is quite generic) to `StrategyDefinition`?  Then this method could be `getStrategyDefinitionMap()` and would be more descriptive of what it returns.

**Comment by pivnicek** [comment]
Maybe for names I'd suggest `Capability` -> `StagingProvider`?  and `Definition` -> `StagingDefinition`?  names are always tough.

File: src/Staging/StrategyFactory.php (hunk 2 above)
Comment (odinuv) [reply src/Staging/StrategyFactory.php:60]: it's in a namespace - InputMapping\Staging\Definition so i'don't like to call it InputMapping\Staging\StagingDefinition :)

**Comment by MiroCillik** [comment]
`Capability` -> `StagingProvider` +1 

`Fullfilment` -> `Operation` | `Use` | `Utilization` maybe? (If I'm getting it right :))

**Comment by pivnicek** [comment]
hmm,  Scope may be ok for Fulfillment,  sorry for being late with that.

**Comment by odinuv** [comment]
I like Scope! :)

File: src/File/Strategy/Local.php [hunk 3]
```diff
@@ -9,18 +9,29 @@ class Local extends AbstractStrategy implements StrategyInterface
...
-        $fs->mkdir($destination);
         parent::downloadFiles($fileConfigurations, $destination);
     }
 
     public function downloadFile($fileInfo, $destinationPath)
     {
         if ($fileInfo['isSliced']) {
-            $this->clientWrapper->getBasicClient()->downloadSlicedFile($fileInfo['id'], $destinationPath);
+            $fs = new Filesystem();
+            $fs->mkdir($this->ensurePathDelimiter($this->dataStorage->getPath()) . $destinationPath);
```
Comment (odinuv) [inline src/File/Strategy/Local.php:19]: destinationPath is a directory

File: src/File/Strategy/Local.php [hunk 4]
```diff
@@ -9,18 +9,29 @@ class Local extends AbstractStrategy implements StrategyInterface
...
+            $this->clientWrapper->getBasicClient()->downloadSlicedFile(
+                $fileInfo['id'],
+                $this->ensurePathDelimiter($this->dataStorage->getPath()) . $destinationPath
+            );
         } else {
-            $this->clientWrapper->getBasicClient()->downloadFile($fileInfo['id'], $destinationPath);
+            $fs = new Filesystem();
+            $fs->mkdir(dirname($this->ensurePathDelimiter($this->dataStorage->getPath()) . $destinationPath));
```
Comment (odinuv) [inline src/File/Strategy/Local.php:26]: destinationPath is a file, use dirname to get the directory to be created

File: tests/Functional/DownloadFilesTest.php [hunk 5]
```diff
@@ -296,33 +299,34 @@ public function testReadFilesLimit()
...
         $finder = new Finder();
-        $finder->files()->in($root . "/download")->notName('*.manifest');
+        $finder->files()->in($this->temp->getTmpFolder() . "/download")->notName('*.manifest');
         self::assertEquals(100, $finder->count());
 
-        $tmpDir = new Temp('file-test');
-        $tmpDir->initRunFolder();
-        $reader = new Reader($this->clientWrapper, new NullLogger(), new NullWorkspaceProvider());
+        $fs = new Filesystem();
+        $fs->remove($this->temp->getTmpFolder());
```
Comment (odinuv) [inline tests/Functional/DownloadFilesTest.php:323]: the temp directory is now bound to the provider, so i have to cleanup the current temp, a new one won't receive the files

File: src/Table/Strategy/AbstractStrategy.php [hunk 6]
```diff
@@ -33,19 +36,31 @@ abstract class AbstractStrategy implements StrategyInterface
...
     }
 
+    protected function ensurePathDelimiter($path)
+    {
+        return $this->ensureNoPathDelimiter($path) . '/';
+    }
+
+    protected function ensureNoPathDelimiter($path)
+    {
+        return rtrim($path, '\\/');
```
Comment (pivnicek) [inline src/Table/Strategy/AbstractStrategy.php:61]: are we sure that the trailing slash will always be escaped?

Comment (odinuv) [reply src/Table/Strategy/AbstractStrategy.php:61]: it's not getting escaped, just trimmed

Comment (pivnicek) [reply src/Table/Strategy/AbstractStrategy.php:61]: right, I just don't understand why the backslashes?

Comment (odinuv) [reply src/Table/Strategy/AbstractStrategy.php:61]: it strips both backslash (which has to be escaped) and forwardslash (which doesn't need to be escaped)

Comment (pivnicek) [reply src/Table/Strategy/AbstractStrategy.php:61]: ah, right, I hadn't thought about the need for trimming a backslash but best to be MS compiant :-D 👍

**Review by pivnicek — APPROVED** [review]
I think it's ok.  docker tests may bring something forward, but I don't see anything crazy here :-)

**Comment by odinuv** [comment]
V kontextu toho, ze to proslo az do runneru https://travis-ci.com/github/keboola/docker-bundle/builds/209620462 tak to povazuju za prijatelny a merguju

**Pull request merged**